datapath: Add eni parent interface to short-cut the FIB lookup

l1b0k opened the pull request:
Introduce THIS_INTERFACE_PARENT_IFINDEX for container endpoints to enable direct packet redirection using bpf_redirect_neigh(). This optimization eliminates the overhead of BPF FIB lookup operations in the datapath.

Flow up #36310

julianwiedmann reviewed (changes requested):
Please limit the optimization to `CT_REPLY` results, so that it matches what we do in the corresponding redirect logic at the `to-netdev` level: https://github.com/cilium/cilium/blob/60f639c555a8ffb56ba1830b35bd1737c7030e28/bpf/lib/nodeport_egress.h#L329

l1b0k committed:
cilium: Add per-endpoint direct routing support

Introduce THIS_INTERFACE_PARENT_IFINDEX for container endpoints
to enable direct packet redirection using bpf_redirect_neigh(). This
optimization short-cut the FIB lookup. Redirect straight to the desired parent interface.
THIS_INTERFACE_PARENT_IFINDEX is an endpoint level config, this can override the DIRECT_ROUTING_DEV_IFINDEX.

Signed-off-by: l1b0k <libokang.dev@gmail.com>

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -1283,9 +1292,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
+#else
 		int oif = 0;
+#endif
+		if (oif > 0 && ct_is_reply4(get_ct_map4(tuple), tuple)) {
```
```suggestion
		if (oif > 0 && ct_status == CT_REPLY) {
```

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -1283,9 +1292,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
+			ret = fib_redirect_v4(ctx, ETH_HLEN, ip4, false, false, ext_err, &oif);
+		}
 
-		ret = fib_redirect_v4(ctx, ETH_HLEN, ip4, false, false, ext_err, &oif);
```
Curious, have you looked at what changes are needed to re-use `fib_redirect_v4()` here? I'd be optimistic that it's possible.

julianwiedmann reviewed (changes requested):
Thank you! Some comments inline. Besides that - have you already looked at how to test these changes?
